pkg/controller/add_cluster.go

```diff
@@ -0,0 +1,167 @@
+	for _, svcCfg := range cls.M3DBCluster.Spec.ServiceConfigurations {
+		if svcCfg.Name == "m3dbnode" {
+			// TODO(PS) replace statefulsets with pods instead
+			if err := c.k8sclient.EnsureStatefulSets(cls.M3DBCluster, svcCfg); err != nil {
+				redStatus.Message = fmt.Sprintf("%s: %v", redStatus.Message, err)
+				if err := c.updateM3DBStatus(cls.M3DBCluster, redStatus); err != nil {
+					return err
+				}
+				return err
+
+			}
+		}
+	}
+	if err := c.EnsureNamespace(cls.M3DBCluster); err != nil {
+		redStatus.Message = fmt.Sprintf("%s: %v", redStatus.Message, err)
+		if err := c.updateM3DBStatus(cls.M3DBCluster, redStatus); err != nil {
+			return err
+		}
+		return err
+	}
+	if err := c.EnsurePlacement(cls.M3DBCluster); err != nil {
+		redStatus.Message = fmt.Sprintf("%s: %v", redStatus.Message, err)
+		if err := c.updateM3DBStatus(cls.M3DBCluster, redStatus); err != nil {
+			return err
+		}
+		return err
+	}
+	status = myspec.M3DBStatus{
+		State:   myspec.GreenState,
+		Message: "m3 db is up and running",
+	}
+	return c.updateM3DBStatus(cls.M3DBCluster, status)
+}
+
+// EnsurePlacement ensures that a placement exists otherwise create one
+func (c *Controller) EnsurePlacement(cluster *myspec.M3DBCluster) error {
+	//get placement
+	_, err := c.placementClient.Get()
+	if err == placement.ErrPlacementNotFound {
+		placementInitRequest := &admin.PlacementInitRequest{
+			NumShards:         cluster.Spec.NumberOfShards,
+			ReplicationFactor: cluster.Spec.ReplicationFactor,
+		}
+		placementDetails, err := c.k8sclient.GetPlacementDetails(cluster)
+		if err != nil {
+			return err
+		}
+		for hostname, zone := range placementDetails {
+			fqdnHostname := fmt.Sprintf("%s.%s", hostname, plc.DefaultServiceName)
+			instance := &placementpb.Instance{
+				Id:             hostname,
+				IsolationGroup: zone,
+				Zone:           plc.DefaultServiceZone,
+				Weight:         100, // TODO(PS) Remove once [PR](https://github.com/m3db/m3/pull/901) is merged
+				Hostname:       fqdnHostname,
+				Endpoint:       fmt.Sprintf("%s:%d", fqdnHostname, _defaultM3DBPort),
+				Port:           _defaultM3DBPort,
+			}
+			placementInitRequest.Instances = append(placementInitRequest.Instances, instance)
+		}
+		if err := c.placementClient.Init(placementInitRequest); err != nil {
+			return err
+		}
+	} else if err != nil {
+		c.logger.Error("failed to apply placement", zap.Error(err))
+		return err
+	}
+	return nil
+}
+
+// EnsureNamespace will retrieve current namespaces to ensure one matches
+// the cluster name or create a new namespace to match the cluster name
+func (c *Controller) EnsureNamespace(cluster *myspec.M3DBCluster) error {
+	//get namespace
+	resp, err := c.namespaceClient.Get()
+	if err != nil {
+		c.logger.Error("failed to get namespace ", zap.Error(err))
+		return err
+	}
+
+	// check if namespace already exist
+	//
+	// TODO(PS) Ensure all namepaces can be validated in a registry
+	namespaces := resp.GetRegistry().GetNamespaces()
```

Comment:
this is what i mean about namespace client returning native types - would allow the logic here to be a bit simpler - you'd be guaranteed the Registry/Metadatas are valid by this point.

Reply:
Fair point and I agree.